File: Leetcode/leetcode-0144-BinaryTreePreorderTraversal-[tree]-[easy]/Solution.iterative2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        
        if (root == null) {
            return result;
        }
        
        stack.push(root);
        while (stack.size() != 0) {
            TreeNode cur = stack.pop();
            result.add(cur.val);
            if (cur.right != null) {
                stack.push(cur.right);
            }
            
            if (cur.left != null) {
                stack.push(cur.left);
            }
        }
        
        return result;
    }
}

File: Leetcode/leetcode-0144-BinaryTreePreorderTraversal-[tree]-[easy]/Solution.iterative2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function (root) {
    let result = [];
    if (root === null) {
        return result;
    }

    let stack = [];
    stack.push(root);
    while (stack.length !== 0) {
        let cur = stack.pop();
        result.push(cur.val);

        if (cur.right !== null) {
            stack.push(cur.right);
        }

        if (cur.left !== null) {
            stack.push(cur.left);
        }
    }

    return result;
};
